File: solutions/lua/palindrome-products/1/palindrome-products.lua
local function isPalindrome(n)
	local s = tostring(n)
	return s == s:reverse()
end

local memo = {}
local function all(min, max)
	assert(min <= max, "min must be <= max")
	if memo[min] and memo[min][max] then return memo[min][max] end
	local retval = {}
	for i = min, max do
		for j = i, max do
			local p = i * j
			if isPalindrome(p) then
				retval[p] = retval[p] or {}
				table.insert(retval[p], { i, j })
			end
		end
	end
	memo[min] = memo[min] or {}
	memo[min][max] = retval
	return retval
end

local function smallest(min, max)
	local r = all(min, max)
	local m = math.huge
	for k in pairs(r) do
		m = math.min(m, k)
	end
	if m == math.huge then return { value = nil, factors = {} } end
	return { value = m, factors = r[m] }
end

local function largest(min, max)
	local r = all(min, max)
	local m = 0
	for k in pairs(r) do
		m = math.max(m, k)
	end
	if m == 0 then return { value = nil, factors = {} } end
	return { value = m, factors = r[m] }
end

return { smallest = smallest, largest = largest }

File: solutions/ruby/palindrome-products/1/palindrome_products.rb
# frozen_string_literal: true

# Pal
Pal = Struct.new(:value, :factors)

# class Palindromes
class Palindromes
  attr_reader :smallest, :largest, :max, :min

  def initialize(max_factor: 0, min_factor: 0)
    @max = max_factor
    @min = min_factor
  end

  def generate
    raise ArgumentError, 'min must be <= max' unless min <= max

    @smallest = @largest = Pal.new(nil, [])

    (min * min).upto(max * max).each do |num|
      next unless pal?(num)

      f = factors(num)
      @smallest = Pal.new(num, f) and break if f.any?
    end
    (max * max).downto(min * min).each do |num|
      next unless pal?(num)

      f = factors(num)
      @largest = Pal.new(num, f) and break if f.any?
    end
  end

  private

  def factors(num)
    (min..Integer.sqrt(num).floor).filter_map do |i|
      if i.positive? && (num % i).zero?
        j = (num / i).floor
        [i, j] if (min..max).cover?(j)
      end
    end
  end

  def pal?(num)
    s = num.to_s
    s == s.reverse
  end
end
